### PR TITLE
Added Focus and Role Items to the Game

### PR DESCRIPTION
**Pull request by JFranklin3000**
##### In this PR
- Added Focus.hbs
- Added role.hbs
- Added initial handlers for roles and advancement

##### Testing
- Validate sheet looks like we want it to.

### CODE REVIEW

File: module/helpers/config.mjs [hunk 1]
```diff
@@ -140,6 +170,17 @@ E20.armorTraits = {
 };
 preLocalize("armorTraits");
 
+// Options for Armor types
+E20.armorTypes = {
+  computerized: "E20.ArmorComputerized",
```
Comment (phildominguez) [inline module/helpers/config.mjs:175]:
```suggestion
  computerized: "E20.ArmorsComputerized",
```

File: module/helpers/config.mjs [hunk 2]
```diff
@@ -230,6 +271,31 @@ E20.skillsByEssence = {
   social: ["animalHandling", "deception", "performance", "persuasion", "streetwise"],
 };
 
+E20.essenceLevels = {
+  level1: "E20.Level1",
+  level1optional: "E20.Level1Additional",
```
Comment (phildominguez) [inline module/helpers/config.mjs:276]: `E20.Level1Additional` not defined

File: module/sheets/actor-sheet.mjs [hunk 3]
```diff
@@ -397,6 +401,8 @@ export class Essence20ActorSheet extends ActorSheet {
     inputs.attr('readonly', isLocked);
     // Don't readonly health and stun values
     html.find('.no-lock').attr('readonly', false);
+    //
+    html.find('.level').change(this._advHandler.onLevelChange(this.actor));
```
Comment (phildominguez) [inline module/sheets/actor-sheet.mjs:405]: I'd move this up with the other handler stuff, like under transform or something. Don't forget the comment!

File: module/sheet-handlers/advancement-handler.mjs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+export class AdvancementHandler {
+
+  /**
```
Comment (phildominguez) [inline module/sheet-handlers/advancement-handler.mjs:3]:
```suggestion
export class AdvancementHandler {
  /**
```

File: module/sheet-handlers/advancement-handler.mjs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  constructor(actorSheet) {
+    this._actorSheet = actorSheet;
+    this._actor = actorSheet.actor;
+  }
+
+  onLevelChange(actor) {
+    console.log(actor);
+  }
+
+}
```
Comment (phildominguez) [inline module/sheet-handlers/advancement-handler.mjs:16]:
```suggestion
  }
}
```

File: module/sheet-handlers/role-handler.mjs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+export class RoleHandler {
+
+  /**
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:3]:
```suggestion
export class RoleHandler {
  /**
```

File: module/sheet-handlers/role-handler.mjs [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  * Constructor
+  * @param {Essence20ActorSheet} actorSheet The actor sheet
+  */
+  constructor(actorSheet) {
+    this._actorSheet = actorSheet;
+    this._actor = actorSheet.actor;
+  }
+
+}
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:12]:
```suggestion
  }
}
```

File: module/helpers/config.mjs [hunk 8]
```diff
@@ -2,6 +2,19 @@ import { preLocalize } from "./utils.mjs";
 
 export const E20 = {};
 
+/************************************************
+ * System Version                               *
+ ***********************************************/
+
+//Game Version
```
Comment (phildominguez) [inline module/helpers/config.mjs:9]:
```suggestion
// Game Version
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/helpers/config.mjs [hunk 9]
```diff
@@ -230,6 +271,31 @@ E20.skillsByEssence = {
   social: ["animalHandling", "deception", "performance", "persuasion", "streetwise"],
 };
 
+E20.essenceLevels = {
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/helpers/config.mjs:274]: I was thinking of it as the level you got the essence boost initially, but yea actorLevels or playerLevels would be better

File: templates/item/sheets/role.hbs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+        <option value="{{type}}">{{name}}</option>
+        {{/each}}
+        {{/select}}
+      </select>
+      {{/inline}}
+      {{/"systems/essence20/templates/item/parts/sheet-field.hbs"}}
+
+      {{#ifNotEquals system.version 'myLittlePony'}}
+
+        {{#ifNotEquals system.version 'transformers'}}
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/item/sheets/role.hbs:25]: Not saying you are wrong, but the ifNotEquals for myLittlePony is removing most of the options. where the transformers one is removing just that one field. but I can take a look.

[the remark this answers is not in the document]
Comment (phildominguez) [reply templates/item/sheets/role.hbs:25]: Ah okay, np

File: templates/item/sheets/focus.hbs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    {{#*inline "item-field-inputs"}}
+    <div class="flex-group-right chip-section">
+    {{#each item.system.skills as |name|}}
+      <span class="chip" name="chip.skills.{{type}}">{{lookup @root.config.originSkills name}}</span>
+    {{/each}}
+    </div>
+    {{/inline}}
+    {{/"systems/essence20/templates/item/parts/sheet-field.hbs"}}
+
+    {{!-- Focus Role --}}
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/item/sheets/focus.hbs:47]: fixed

File: module/helpers/config.mjs (hunk 2 above)
Comment (JFranklin3000) [reply module/helpers/config.mjs:276]: fixed

File: module/sheets/actor-sheet.mjs (hunk 3 above)
Comment (JFranklin3000) [reply module/sheets/actor-sheet.mjs:405]: fixed

File: template.json [hunk 12]
```diff
@@ -829,6 +841,39 @@
...
+        "qualified": "",
+        "trained":""
+      },
+      "essences": {
+        "smarts":  "",
+        "social":  "",
+        "speed":  "",
+        "strength":  ""
+      },
+      "perks": {
```
Comment (phildominguez) [inline template.json:859]: Should this be perkIds?

File: template.json [hunk 13]
```diff
@@ -829,6 +841,39 @@
...
+          "increase": 0,
+          "levels": "",
+          "starting": 0
+        }
+      },
+      "skills": [],
+      "version": "",
+      "weapons": {
+        "qualified": "",
+        "trained":""
```
Comment (phildominguez) [inline template.json:874]: Should these be lists?

File: template.json [hunk 14]
```diff
@@ -829,6 +841,39 @@
...
       "usesPer": null
     },
+    "role": {
+      "templates": [
+        "item",
+        "itemDescription"
+      ],
+      "armors": {
+        "qualified": "",
+        "trained":""
```
Comment (phildominguez) [inline template.json:851]: Should these be lists?

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: template.json (hunk 12 above)
Comment (JFranklin3000) [reply template.json:859]: I would say no in this case right now it is an array of levels.

File: template.json (hunk 13 above)
Comment (JFranklin3000) [reply template.json:874]: yes

File: template.json (hunk 14 above)
Comment (JFranklin3000) [reply template.json:851]: yes

File: template.json [hunk 15]
```diff
@@ -829,6 +841,38 @@
...
+    "role": {
+      "templates": [
+        "item",
+        "itemDescription"
+      ],
+      "armors": {
+        "qualified": [],
+        "trained": []
+      },
+      "essences": {
```
Comment (phildominguez) [inline template.json:853]: Maybe `essenceLevels`?

File: template.json [hunk 16]
```diff
@@ -829,6 +841,38 @@
...
+        "smarts":  "",
+        "social":  "",
+        "speed":  "",
+        "strength":  ""
```
Comment (phildominguez) [inline template.json:857]: Lists I think

File: template.json [hunk 17]
```diff
@@ -829,6 +841,38 @@
...
+      },
+      "perks": {
+        "general": []
+      },
+      "powers": {
+        "personal": {
+          "increase": 0,
+          "levels": "",
```
Comment (phildominguez) [inline template.json:865]: This name sounds like it should be a list?

File: template.json [hunk 18]
```diff
@@ -829,6 +841,38 @@
...
+          "starting": 0
+        }
+      },
+      "skills": [],
+      "version": "",
+      "weapons": {
+        "qualified": [],
+        "trained":[]
```
Comment (phildominguez) [inline template.json:873]:
```suggestion
        "trained": []
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: template.json (hunk 18 above)
Comment (JFranklin3000) [reply template.json:873]: fixed

File: template.json (hunk 17 above)
Comment (JFranklin3000) [reply template.json:865]: fixed

File: template.json (hunk 16 above)
Comment (JFranklin3000) [reply template.json:857]: fixed

File: template.json (hunk 15 above)
Comment (JFranklin3000) [reply template.json:853]: changed

**Review by phildominguez — APPROVED** [review]

**Pull request merged**